### PR TITLE
feat(core): Targeted message encryption (WEBFOUND-17)

### PR DESCRIPTION
**Pull request by ffflorian**
With this PR a message can now be encrypted for certain users only.

## Pull Request Checklist

- [ ] My code is covered by tests
- [x] I will [merge the PR as breaking change](https://github.com/wireapp/wire-web-packages/wiki/Releases#create-major-release), if the API contract changes

### CODE REVIEW

File: packages/core/src/main/conversation/ConversationService.ts [hunk 1]
```diff
@@ -144,6 +146,28 @@ class ConversationService {
...
+  ): Promise<PayloadBundleOutgoing> {
+    const confirmationContent = payloadBundle.content as ConfirmationContent;
+
+    const confirmationObject: PayloadBundleOutgoingUnsent = {
+      ...payloadBundle,
+      type: GenericMessageType.CONFIRMATION,
+    };
+
+    await this.send(conversationId, confirmationObject);
+    await this.deleteMessageEveryone(conversationId, confirmationContent.confirmMessageId);
```
Comment (atomrc) [inline packages/core/src/main/conversation/ConversationService.ts]: Humm there seems to be a flaw in the process here. 
Basically what this code means is that the first that receives the ephemeral message will actually try to delete it for everyone.

In the webapp, what we do is sending this deletion message (in the same conversation) but only targeted (/encrypted) for the person who actually sent the message. 

(see https://github.com/wireapp/wire-webapp/blob/e1357df97249e328cb24e6977444af7417f0513c/app/script/conversation/ConversationRepository.js#L2551)

As you can see, in the webapp, we have a parameter for the `delete_message_everyone` where we can specify a bunch of user ids to send the message to.

Maybe it's the flow to use in that case, maybe we should invent something new, but we should not send the `delete` message to everyone in the convo. 

ping @gregor @bennyn

Comment (ffflorian) [reply packages/core/src/main/conversation/ConversationService.ts]: That's technically correct, but we never implemented encrypting messages for certain people. So for now we have to send the `MessageDelete` to everyone in the group. That's a bit of overhead but every client needs to be able to handle this anyway.

File: packages/core/src/main/conversation/ConversationService.ts [hunk 2]
```diff
@@ -159,6 +161,38 @@ class ConversationService {
...
+    payloadBundle: PayloadBundleOutgoingUnsent
+  ): Promise<PayloadBundleOutgoing> {
+    const confirmationContent = payloadBundle.content as ConfirmationContent;
+
+    const confirmationObject: PayloadBundleOutgoingUnsent = {
+      ...payloadBundle,
+      type: PayloadBundleType.CONFIRMATION,
+    };
+
+    await this.send(conversationId, confirmationObject);
```
Comment (atomrc) [inline packages/core/src/main/conversation/ConversationService.ts]: Here the method sends the `confirmation` and the `ephemeralConfirmation` at the same time.
The timeline should be this way:
- send `confirmation` right away ;
- wait until timer is out ;
- send `ephemeralConfirmation`.

Also, if you use `timeout` for that, you are going to run into issues if you have timers > a few days (don't remember exactly the number, something like 20)

File: packages/core/src/demo/echo.js [hunk 3]
```diff
@@ -41,6 +41,18 @@ const messageEchoCache = {};
...
 
+  const sendConfirmation = async (messageId, conversationId, messageTimer, userId) => {
+    const confirmationPayload = account.service.conversation.createConfirmation(messageId);
+    await account.service.conversation.send(conversationId, confirmationPayload);
+
+    if (messageTimer && userId) {
+      setTimeout(
+        () => account.service.conversation.deleteMessageEveryone(conversationId, messageId, [userId]),
+        messageTimer
+      );
```
Comment (gregor) [inline packages/core/src/demo/echo.js]: Still don't get why we would send this confirmation on messageTimer timeout even in the demo case.

File: packages/core/src/main/Account.ts [hunk 4]
```diff
@@ -251,13 +251,12 @@ class Account extends EventEmitter {
...
-          const expireAfterMillis = genericMessage.ephemeral.expireAfterMillis;
-          unwrappedMessage.messageTimer =
-            typeof expireAfterMillis === 'number' ? expireAfterMillis : (expireAfterMillis as Long).toNumber();
-        }
+        unwrappedMessage.id = genericMessage.messageId;
+        const expireAfterMillis = genericMessage.ephemeral.expireAfterMillis;
+        unwrappedMessage.messageTimer =
+          typeof expireAfterMillis === 'number' ? expireAfterMillis : (expireAfterMillis as Long).toNumber();
         return unwrappedMessage;
       } else {
```
Comment (gregor) [inline packages/core/src/main/Account.ts]: No need for else after return.

File: packages/core/src/demo/echo.js (hunk 3 above)
Comment (ffflorian) [reply packages/core/src/demo/echo.js]: But we send a confirmation in [line 46](https://github.com/wireapp/wire-web-packages/pull/939/files#diff-d497294ce653e28d25a71b3bf9dfb025R46).

File: packages/core/src/demo/echo.js [hunk 5]
```diff
@@ -41,6 +41,15 @@ const messageEchoCache = {};
   const apiClient = new APIClient({store: engine, urls: backend});
   const account = new Account(apiClient);
 
+  const sendConfirmation = async (messageId, conversationId, messageTimer, userId) => {
```
Comment (gregor) [inline packages/core/src/demo/echo.js]: I still don't think this is the appropriate name. Confirmation is used within the scope of the app to confirm the delivery of the message.

What you are doing here is a message deletion after ephemeral timer run out. Confirmation is not the correct term to be used.

File: packages/core/src/demo/echo.js [hunk 6]
```diff
@@ -41,36 +41,62 @@ const messageEchoCache = {};
...
-      `Message "${messageId}" in "${conversationId}" from "${from}":`,
-      content.text,
+      `Receiving: "${type}" ("${messageId}") in "${conversationId}" from "${from}":`,
+      content,
       messageTimer ? `(ephemeral message, ${messageTimer} ms timeout)` : ''
     );
 
-    const confirmationPayload = account.service.conversation.createConfirmation(messageId);
-    await account.service.conversation.send(conversationId, confirmationPayload);
+    if (messageId) {
```
Comment (gregor) [inline packages/core/src/demo/echo.js]: In the long run a message ID should not be the trigger of for a confirmation. There are specific content messages that need to be confirmed. Call messages won't be for example.

**Pull request merged**